File: apps/api/app/core/finding_state.py
# SPDX-FileCopyrightText: 2026 Virantis
# SPDX-License-Identifier: LicenseRef-Vooda-Community-1.0
"""Single source of truth for what a finding's ``classification`` MEANS.

Before this module every caller hand-rolled its own tuple of
classifications, and they disagreed. ``LIKELY_FALSE_POSITIVE`` counted as
CLOSED in the metrics router and as UNDECIDED in the occurrence
propagator — a contradiction covering ~90% of all findings, which meant
the dashboard reported an unconfirmed AI guess as settled work.

The rule this module encodes: **only a decision closes a finding.** An
AI verdict is an opinion. ``LIKELY_*`` is therefore open — the AI's
opinion is carried by ``ai_confidence``, not by the closure state. That
is also why GitHub, GitGuardian and GitLab all require an explicit
resolution reason to leave the open state.

Adding a value to ``Classification`` without adding it to exactly one of
``CLOSED`` / ``OPEN`` raises at import time (see ``_assert_total``), so
the taxonomy cannot silently drift again.
"""

from typing import Iterable

from apps.api.app.models.finding import Classification

# ── The primary axis: is there still work to do? ─────────────────────

#: Settled. Somebody decided, or the credential was provably neutralised.
#: Nothing here is the product of an AI verdict alone.
CLOSED: tuple[Classification, ...] = (
    Classification.CONFIRMED_FALSE_POSITIVE,
    Classification.TEST_CREDENTIAL,
    Classification.ACCEPTED_RISK,
    Classification.ROTATED,
    Classification.RESOLVED_FILE_DELETED,
    Classification.RESOLVED_ITEM_DELETED,
    Classification.RESOLVED_REPO_REMOVED,
    Classification.RESOLVED_SOURCE_REMOVED,
)

#: Open work. Note CONFIRMED_TRUE_POSITIVE is open: a confirmed real
#: secret is the most open thing in the system until it is rotated.
OPEN: tuple[Classification, ...] = (
    Classification.NEEDS_REVIEW,
    Classification.NOT_ENOUGH_EVIDENCE,
    Classification.LIKELY_TRUE_POSITIVE,
    Classification.LIKELY_FALSE_POSITIVE,
    Classification.CONFIRMED_TRUE_POSITIVE,
)

# ── Secondary axes. Orthogonal to open/closed; never a substitute. ───

#: Nobody has ruled on these yet, so a verdict may propagate into them.
#: A superset of AI_ADVISORY: it also covers "no opinion at all".
UNDECIDED: tuple[Classification, ...] = (
    Classification.NEEDS_REVIEW,
    Classification.NOT_ENOUGH_EVIDENCE,
    Classification.LIKELY_TRUE_POSITIVE,
    Classification.LIKELY_FALSE_POSITIVE,
)

#: An AI opinion with no human behind it. Safe to present as a
#: low-priority bucket; never safe to present as resolved.
AI_ADVISORY: tuple[Classification, ...] = (
    Classification.LIKELY_TRUE_POSITIVE,
    Classification.LIKELY_FALSE_POSITIVE,
)

#: Open, but the AI judged it not a real exposure. These are what a
#: findings list should collapse by default — hidden from the front
#: page, still counted as open.
AI_LOW_RISK: tuple[Classification, ...] = (
    Classification.LIKELY_FALSE_POSITIVE,
)

#: A human decided. Requires classification_provenance.
CONFIRMED: tuple[Classification, ...] = (
    Classification.CONFIRMED_TRUE_POSITIVE,
    Classification.CONFIRMED_FALSE_POSITIVE,
)

#: "Judged not a real exposure", at any confidence. For FP-rate
#: reporting only — do NOT use as a closure test.
FALSE_POSITIVE_VERDICTS: tuple[Classification, ...] = (
    Classification.LIKELY_FALSE_POSITIVE,
    Classification.CONFIRMED_FALSE_POSITIVE,
)

#: "Judged a real exposure", at any confidence.
TRUE_POSITIVE_VERDICTS: tuple[Classification, ...] = (
    Classification.LIKELY_TRUE_POSITIVE,
    Classification.CONFIRMED_TRUE_POSITIVE,
)

#: Carries an AI verdict of any kind. Used to answer "has triage run
#: on this finding yet?" — not "is it settled".
ANY_VERDICT: tuple[Classification, ...] = (
    FALSE_POSITIVE_VERDICTS + TRUE_POSITIVE_VERDICTS
)

#: The leak was actually cleaned up — the only states MTTR may count.
#: Admin deletions (RESOLVED_REPO_REMOVED / RESOLVED_SOURCE_REMOVED) are
#: deliberately absent: they mean "Vooda lost visibility", not "fixed".
#: Counting them would let deleting a repository read as an instant fix.
REMEDIATED: tuple[Classification, ...] = (
    Classification.ROTATED,
    Classification.RESOLVED_FILE_DELETED,
    Classification.RESOLVED_ITEM_DELETED,
)


def _assert_total() -> None:
    """Every classification must be exactly one of CLOSED / OPEN."""
    covered = set(CLOSED) | set(OPEN)
    missing = set(Classification) - covered
    if missing:
        raise RuntimeError(
            "finding_state: unclassified Classification value(s): "
            f"{sorted(m.value for m in missing)}. Add each to CLOSED or "
            "OPEN in apps/api/app/core/finding_state.py."
        )
    overlap = set(CLOSED) & set(OPEN)
    if overlap:
        raise RuntimeError(
            "finding_state: value(s) in both CLOSED and OPEN: "
            f"{sorted(o.value for o in overlap)}."
        )


_assert_total()


# ── SQLAlchemy helpers. Use these instead of inlining a set. ─────────

def is_open(model) -> "object":
    """Clause: this finding still represents work to do."""
    return model.classification.notin_(CLOSED)


def is_closed(model) -> "object":
    """Clause: this finding is settled."""
    return model.classification.in_(CLOSED)


def in_states(model, states: Iterable[Classification]) -> "object":
    """Clause: classification is one of ``states``."""
    return model.classification.in_(tuple(states))

File: apps/web/src/lib/findingState.ts
// SPDX-FileCopyrightText: 2026 Virantis
// SPDX-License-Identifier: LicenseRef-Vooda-Community-1.0
//
// Client-side twin of apps/api/app/core/finding_state.py.
//
// The backend's vocabulary was consolidated first; the dashboard, the
// finding panel and the incident drawer each still carried their own
// copy of "what counts as settled", and all three counted
// likely_false_positive as settled — telling the operator "no action
// needed" on the strength of an AI guess nobody had reviewed.
//
// The rule, same as the backend: only a decision closes a finding. An
// AI verdict is an opinion, shown as such.

/** Settled — somebody decided, or the credential was neutralised. */
export const CLOSED = [
  "confirmed_false_positive",
  "test_credential",
  "accepted_risk",
  "rotated",
  "revoked",
  "resolved_file_deleted",
  "resolved_item_deleted",
  "resolved_repo_removed",
  "resolved_source_removed",
] as const;

/** An AI verdict with nobody behind it. Open, but de-prioritised. */
export const AI_LOW_RISK = ["likely_false_positive"] as const;

/** AI opinion of any kind — advisory, never a resolution. */
export const AI_ADVISORY = ["likely_true_positive", "likely_false_positive"] as const;

const norm = (c?: string | null) => (c || "").toLowerCase();

/** Settled. Mirrors is_closed() server-side, incl. legacy resolved_*. */
export function isClosed(cls?: string | null): boolean {
  const c = norm(cls);
  return (CLOSED as readonly string[]).includes(c) || c.startsWith("resolved");
}

/** Still work to do. */
export function isOpen(cls?: string | null): boolean {
  return !isClosed(cls);
}

/** Open, but the model judged it not a real exposure. */
export function isAiLowRisk(cls?: string | null): boolean {
  return (AI_LOW_RISK as readonly string[]).includes(norm(cls));
}

/** Carries an AI verdict rather than a human decision. */
export function isAiAdvisory(cls?: string | null): boolean {
  return (AI_ADVISORY as readonly string[]).includes(norm(cls));
}

/**
 * Display label. The AI verdicts are deliberately prefixed: calling
 * likely_false_positive "False Positive" asserts a conclusion the
 * system has not reached, and reads identically to the human-confirmed
 * verdict beside it in the same list.
 */
export function classificationLabel(cls?: string | null): string {
  switch (norm(cls)) {
    case "needs_review": return "Needs Review";
    case "not_enough_evidence": return "Needs Review";
    case "likely_true_positive": return "AI: likely real";
    case "likely_false_positive": return "AI: likely not a secret";
    case "confirmed_true_positive": return "Confirmed TP";
    case "confirmed_false_positive": return "Confirmed FP";
    case "accepted_risk": return "Accepted Risk";
    case "test_credential": return "Test Credential";
    case "rotated":
    case "revoked":
    case "resolved": return "Rotated / Revoked";
    default: return norm(cls).replace(/_/g, " ");
  }
}


/** Human label for the lifecycle pair. Prefers the new columns and
 *  falls back to the legacy classification for any stale payload. */
export function statusLabel(obj: any): string {
  const st = (obj?.status || "").toLowerCase();
  const rs = (obj?.resolution_reason || "").toLowerCase();
  const REASONS: Record<string, string> = {
    rotated: "Rotated",
    revoked: "Revoked",
    provider_disabled: "Provider disabled",
    false_positive: "False positive",
    test_credential: "Test credential",
    acceptable_risk: "Acceptable risk",
    mitigating_control: "Mitigating control",
    no_longer_present: "No longer present",
  };
  if (st === "resolved" || st === "dismissed") {
    const head = st === "resolved" ? "Resolved" : "Dismissed";
    return rs ? `${head} — ${REASONS[rs] || rs.replace(/_/g, " ")}` : head;
  }
  if (st === "triaging") return "Triaging";
  if (st === "open") {
    // An open finding shows the model's opinion, clearly marked as one.
    const v = (obj?.ai_verdict || "").toLowerCase();
    if (v === "likely_fp") return "Open — AI: likely not a secret";
    if (v === "likely_tp") return "Open — AI: likely real";
    return "Open";
  }
  return classificationLabel(obj?.classification);
}
